designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,312 @@
+
+### Q: How should allowed disruptions be calculated in a multi-reason world? 
+The calculation of allowed disruptions in a system with multiple disruption reasons has become more intricate following the introduction of Disruption Budget Reasons. Previously, the formula was straightforward:
+
+AllowedDisruptions = minNodeCountOfAnActiveBudget - unhealthyNodes - totalDisruptingNodes.
+
+When calculating by reason, two potential equations emerge:
+1. AllowedDisruptionByReason = minNodeCountOfActiveBudget[reason] - unhealthyNodes - totalDisruptingNodes[reason]
+2. AllowedDisruptionByReason = minNodeCountOfActiveBudget[reason] - unhealthyNodes - totalDisruptingNodes
+The second equation is the one we should opt into.
+
+Take this budget as an example
+```yaml
+budgets: 
+  - nodes: 15 
+    reasons: [drifted, underutilized]
+    schedule: "* * * * *"
+  - nodes: 10
+    reasons: [drifted]
+    schedule: "* * * * *"
+  - nodes: 5 
+    schedule: "* * * * *" 
+```
+First, calculate minNodeCountOfActiveBudget. Given there are two budgets for drifted, we select the lower number:
+
+```
+minNodeCountOfActiveBudget = {
+  drifted: 10 
+  underutilized: 15 
+  Default: 5
+}
+```
+
+In the first equation, tracking the number of nodes disrupted by each reason is required. For example:
+```
+disrupting = {
+  drifted: 3 
+  underutilized: 6
+  Expriation: 3
+  empty: 2
+}
+```
+
+Assuming zero unhealthy nodes:
+**First Equation**
+- drifted = 10 - 3: Disruption allowed as drifted bucket isn't full.
+- underutilized = 15 - 6: Disruption allowed since 15 - 6 > 0.
+- Default = 5 - expired - empty = 0: No disruption allowed for reasons other than consolidation and drift.
+
+**Second Equation**
+- Total Disrupting = 3 + 6 + 5 = 14.
+- drifted = 10 - 14: No disruption allowed due to drift.
+- underutilized = 15 - 14: Disruption allowed for at least one node due to consolidation.
+- Default: 5 - 14: No disruption allowed for other methods.
+
+#### Decision
+The second equation simplifies reasoning for cluster operators. A node can be tagged for multiple disruption reasons (e.g., empty, drifted, underutilized), making it challenging to precisely calculate in-flight disruptions. The second equation streamlines this process by providing a clearer view of the overall disruption impact.
+### Q: How should we handle an unspecified default reason? 
+```yaml
+budgets: 
+  - nodes: 10
+    reasons: [drift, underutilized]
+    schedule: "* * * * *"
+  - nodes: 5 
+    reasons: [empty] 
+    schedule: "* * * * *" 
+```
+In the case of a budget like above, default is undefined. Should karpenter assume the user doesn't want to disrupt any other reasons? Or should we assume that if a default is unspecified, they want us to disrupt anyway?  
+The intuitive options if there is no active default budget is to allow disruption of either 0 or total number of nodes(meaning unbounded disruption).
```

Comment:
The other point of concern here is that once you are specifying budgets with reasons, we can't default you to 10% like we do generally when you don't specify a budget. I think this is fine, considering if you are getting this into the weeds with your budgets, you should know that you need to set a budget that covers the cases that you didn't explicitly state, but it's worth calling out.

In either case, unbounded strikes me as the right behavior.

Reply:
This is the expected behavior today as well. If you don't have a budget with a schedule matching the current time, then it'll be unbounded.

Reply:
@njtran When i test this on the current karpenter core v0.34.1, seems that not specifying a budget, i get 10% budget applied by default. 

Applying this nodepool with no budget specified https://github.com/Azure/karpenter-provider-azure/blob/main/examples/v1beta1/general-purpose.yaml, we get 10%, and disruption is blocked due to this budget. 

```
~/dev/karp/karpenter-provider-azure (bsoghigian/v0.34-clean*) » k get nodepools.karpenter.sh general-purpose -o yaml                          sillygoose@Bryces-MacBook-Pro
apiVersion: karpenter.sh/v1beta1
kind: NodePool
.....
spec:
  disruption:
    budgets:
    - nodes: 10%
    consolidationPolicy: WhenUnderutilized

```